contracts/DAOSignApp.sol

```diff
@@ -0,0 +1,204 @@
+
+    function storeProofOfAuthority(SignedProofOfAuthority memory data) external;
+
+    function storeProofOfSignature(SignedProofOfSignature memory data) external;
+
+    function storeProofOfAgreement(SignedProofOfAgreement memory data) external;
+}
+
+contract DAOSignApp is DAOSignEIP712, DAOSignBaseApp {
+    mapping(string => SignedProofOfAuthority) poaus;
+    mapping(string => SignedProofOfSignature) posis;
+    mapping(string => SignedProofOfAgreement) poags;
+    mapping(string => address) proof2signer;
+    mapping(string => mapping(address => uint256)) poauSignersIdx;
+
+    constructor() {
+        domain.name = 'daosign';
+        domain.version = '0.1.0';
+        DOMAIN_HASH = hash(domain);
+        initEIP712Types();
+    }
+
+    function memcmp(bytes memory a, bytes memory b) internal pure returns (bool) {
+        return (a.length == b.length) && (keccak256(a) == keccak256(b));
+    }
+
+    function strcmp(string memory a, string memory b) internal pure returns (bool) {
+        return memcmp(bytes(a), bytes(b));
+    }
+
+    function validate(SignedProofOfAuthority memory data) internal pure returns (bool) {
+        require(bytes(data.proofCID).length == IPFS_CID_LENGHT, 'Invalid proof CID');
+        require(strcmp(data.message.app, 'daosign'), 'Invalid struct name');
+        require(strcmp(data.message.name, 'Proof-of-Authority'), 'Invalid struct name');
+        require(bytes(data.message.agreementCID).length == IPFS_CID_LENGHT, 'Invalid CID length');
+        for (uint256 i = 0; i < data.message.signers.length; i++) {
+            require(data.message.signers[i].addr != address(0), 'Invalid signer');
+        }
+        return true;
+    }
+
+    function validate(SignedProofOfSignature memory data) internal view returns (bool) {
+        require(bytes(data.proofCID).length == IPFS_CID_LENGHT, 'Invalid proof CID');
+        require(strcmp(data.message.app, 'daosign'), 'Invalid struct name');
+        require(strcmp(data.message.name, 'Proof-of-Signature'), 'Invalid struct name');
+
+        uint i = poauSignersIdx[data.message.agreementCID][data.message.signer];
+        require(
+            poaus[data.message.agreementCID].message.signers[i].addr == data.message.signer,
+            'Invalid signer'
+        );
+
+        return true;
+    }
+
+    function validate(SignedProofOfAgreement memory data) internal view returns (bool) {
+        require(bytes(data.proofCID).length == IPFS_CID_LENGHT, 'Invalid proof CID');
+        require(strcmp(data.message.app, 'daosign'), 'Invalid struct name');
+        require(
+            strcmp(poaus[data.message.agreementCID].message.name, 'Proof-of-Authority'),
+            'Invalid agreementCID'
+        );
+        require(
+            poaus[data.message.agreementCID].message.signers.length ==
+                data.message.signatureCIDs.length,
+            'Invalid sign proofs'
+        );
+        for (uint i = 0; i < data.message.signatureCIDs.length; i++) {
```

Comment:
The order of signers in Proof-of-Authority and in Proof-of-Agreement may not be the same, so it's better not to rely on that

Reply:
good catch, we need to sort it first

Reply:
it uses index checking, please take a closer look at the algorithm

Reply:
@n0cte @kushkamisha can we check in tests?